cmd/pinniped/cmd/kubeconfig_test.go

```diff
@@ -2341,6 +2472,244 @@ func TestGetKubeconfig(t *testing.T) {
 					base64.StdEncoding.EncodeToString([]byte(issuerCABundle)))
 			},
 		},
+		{
+			name: "supervisor upstream IDP discovery when both name and type are specified but flow is not and a matching IDP is found",
+			args: func(issuerCABundle string, issuerURL string) []string {
+				f := testutil.WriteStringToTempFile(t, "testca-*.pem", issuerCABundle)
+				return []string{
+					"--kubeconfig", "./testdata/kubeconfig.yaml",
+					"--skip-validation",
+					"--no-concierge",
+					"--oidc-issuer", issuerURL,
+					"--oidc-ca-bundle", f.Name(),
+					"--upstream-identity-provider-name", "some-ldap-idp",
+					"--upstream-identity-provider-type", "ldap",
+				}
+			},
+			oidcDiscoveryResponse: happyOIDCDiscoveryResponse,
+			idpsDiscoveryResponse: here.Docf(`{
+				"pinniped_identity_providers": [
+					{"name": "some-ldap-idp", "type": "ldap"},
+					{"name": "some-oidc-idp", "type": "oidc"},
+					{"name": "some-other-oidc-idp", "type": "oidc"}
+				]
+			}`),
+			wantStdout: func(issuerCABundle string, issuerURL string) string {
+				return here.Docf(`
+					apiVersion: v1
+					clusters:
+					- cluster:
+						certificate-authority-data: ZmFrZS1jZXJ0aWZpY2F0ZS1hdXRob3JpdHktZGF0YS12YWx1ZQ==
+						server: https://fake-server-url-value
+					  name: kind-cluster-pinniped
+					contexts:
+					- context:
+						cluster: kind-cluster-pinniped
+						user: kind-user-pinniped
+					  name: kind-context-pinniped
+					current-context: kind-context-pinniped
+					kind: Config
+					preferences: {}
+					users:
+					- name: kind-user-pinniped
+					  user:
+						exec:
+						  apiVersion: client.authentication.k8s.io/v1beta1
+						  args:
+						  - login
+						  - oidc
+						  - --issuer=%s
+						  - --client-id=pinniped-cli
+						  - --scopes=offline_access,openid,pinniped:request-audience
+						  - --ca-bundle-data=%s
+						  - --upstream-identity-provider-name=some-ldap-idp
+						  - --upstream-identity-provider-type=ldap
+						  command: '.../path/to/pinniped'
+						  env: []
+						  provideClusterInfo: true
+					`,
+					issuerURL,
+					base64.StdEncoding.EncodeToString([]byte(issuerCABundle)))
+			},
+		},
+		{
+			name: "supervisor upstream IDP discovery when flow is specified and no flows were returned by discovery uses the specified flow",
```

Comment:
Err, this seems weird?

Reply:
Yes, it's a little odd. This is the case where the admin user is running a new version of `pinniped get kubeconfig` against an old version of the Pinniped Supervisor, and they have explicitly specified `--upstream-identity-provider-flow` for some reason even though it was not required. It's a corner case. Since they expended the effort to supply a value for the flag, and since we have no way to validate the value using the old Supervisor's discovery response, we can just pass it through to the user's kubeconfig. If they specify something that the end user's `pinniped login oidc` command does not understand, then that command will print a nice error message. This avoids having the `pinniped get kubeconfig` command need to treat the flow names as anything other than opaque strings.